File: HashMapAndHeap/BasicHM.java
import java.util.HashMap;
import java.util.ArrayList;

public class BasicHM{

public static void test_01()
{
    HashMap<String, Integer> map = new HashMap<>();
    map.put("USA",1000);
    map.put("IND",10000);
    map.put("NEP",90);
    map.put("usa",9000);

    System.out.println(map); // direct map print kr skte hai

    map.put("USA",10022);  // USA is already present so uski value update ho jaayegi
    System.out.println(map);

    System.out.println(map.get("Usa")); // Usa not present so null will be printed

    for(String s : map.keySet()) // .keySet returns an array of keys  
        System.out.println(s + " -> " + map.get(s));

}

//Frequency map of a given string
public static void test_02(String str)
{
    HashMap<Character,Integer> map = new HashMap<>();
    for(int i = 0; i < str.length();i++)
    {
        char ch = str.charAt(i);
        if(!map.containsKey(ch))
            map.put(ch,1);
        else
            map.put(ch,map.get(ch) + 1);
    }

    //OR BETTER SYNTAX ->
    // for(int i = 0; i < str.length(); i++)
    // {
    //     char ch = str.charAt(i);
    //     map.put(ch,map.getOrDefault(ch,0) + 1);
    // }

    System.out.println(map);
}

//Storing indexes of all characters occuring in the string
public static void test_03(String str)
{
    HashMap<Character,ArrayList<Integer>> map = new HashMap<>();
    for(int i = 0; i < str.length(); i++)
    {
        char ch = str.charAt(i);
        map.putIfAbsent(ch,new ArrayList<>()); // pehli baar koi char aaya to uske corresponding new AL add
        map.get(ch).add(i); 
    }

    System.out.println(map);
}

public static void main(String[] args)
{
    // test_01();
    // test_02("abcbaadbdisnduhdej");
    test_03("abcbaadbdisnduhdej");
}
}

File: HashMapAndHeap/BasicHM.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

void test_01()
{
    unordered_map<string,int> map;
    map["USA"] = 1000;
    map["IND"] = 10000;   // similar to using vector
    map["NEP"] = 90;
    map["usa"] = 9000;

    cout << map["USA"] << endl;
    // cout << map["BAN"] << endl; //by default if key not present in UM then cpp adds it into UM with value 0 so 0 print ho jaayega and BAN UM mei add ho jaayega
    
    //to prevent this->
    if(map.find("BAN") != map.end())
        cout << map["BAN"] << endl; // jab map mei mile tbhi get krna 

    // Direct map print nhi kr skte cpp mei -> we use pair of key vs value to iterate over it
    for(pair<string,int> p : map)
        cout << p.first << " -> " << p.second << endl;
}

//Frequency map of a given string
void test_02(string str)
{
    unordered_map<char,int> map;
    for(char ch : str)
    {
        map[ch]++;  // map[ch] krne se khud hi map mei add ho jaata hai ch with 0 value
    }

    for(pair<char,int> p : map)
    {
        cout << p.first << " -> " << p.second << endl;
    }
}

//Indexes of all characters occuring in a string
void test_03(string str)
{
    unordered_map<char,vector<int>> map;
    for(int i = 0; i < str.length(); i++)
    {
        char ch = str[i];
        map[ch].push_back(i); // map[ch] krte hi already empty vector add hogya ch ke against
    }

    //printing the map
    for(auto &key : map)  // using auto for better syntax -> auto automatically req pair le aayega key mei
    {
        char k = key.first;
        vector<int> val = key.second;

        cout << k << " -> " ;
        for(int ele : val)
            cout << ele << ", ";
        cout << endl;

    }
}

int main()
{
    // test_01();
    // test_02("ababnabnajijdidd");
    test_03("ababnabnajijdidd");
}